### PR TITLE
allow to resume sequence upload

### PR DESCRIPTION
**Pull request by richlv**
- if SEQUENCE file is found, read the UUID from it
- if the file is not found, generate and UUID & write it to that file

### CODE REVIEW

**Comment by mprins** [comment]
People are bound to forget about this sequence file, you should add a parameter to use it instead of defaulting to resume.

**Comment by richlv** [comment]
@mprins, could you provide some usecase where forgetting about that file could be harmful ?
for me, the workflow would be uploading files in some directory, then possibly cancelling that and just running the upload again, resuming the sequence. when would one want to ignore the sequence file, if present ?

**Comment by mprins** [comment]
as far as I can see the SEQUENCE file is written in the working directory and is never deleted

**Comment by richlv** [comment]
i think i got what you meant and i attempted to change it so that the file is now written to the directory we upload. i only ever uploaded current directory :)
please note that i have no idea what i'm doing (first time i try to do anything in python), and git is mighty confusing...

**Comment by richlv** [comment]
heya - anybody from mapillary with an insight on what sucks and should be improved in this pull request ? :)

**Comment by ybkuang** [comment]
@richlv We will take a good look at this after the holidays. Sorry for the delay. Merry Christmas!

**Comment by richlv** [comment]
thanks for the response - i've tested it quite a bit myself and it functions really nicely so far :)

**Comment by richlv** [comment]
hei-hou. hope the holidays were great for you all ;)

**Comment by Skippern** [comment]
This will be great, this is a feature I have been missing

**Comment by richlv** [comment]
thanks @ybkuang, i tried to use a variable for the file location as early as possible. if the changes went in the pull request properly, does this approach look ok ?

**Comment by ybkuang** [comment]
@richlv Yes, it looks good.

**Comment by Skippern** [comment]
Any chance this will be implemented in the near future?

**Comment by ybkuang** [comment]
@Skippern I will take a second look at this soon. Sorry for the delay.

**Comment by ToeBee** [comment]
Oh... I kind of assumed the script did this already. Considering it uploads in seemingly random order, any interruption during an hour long transfer means an instant disaster requiring throwing away all the uploaded pictures and starting over.

**Comment by Skippern** [comment]
This would solve my issue with mapillary/mapillary_issues#2081

File: python/upload_with_authentication.py [hunk 1]
```diff
@@ -72,16 +72,16 @@
...
         print("You are missing one of the environment variables MAPILLARY_USERNAME, MAPILLARY_PERMISSION_HASH or MAPILLARY_SIGNATURE_HASH. These are required.")
         sys.exit()
-
-    if os.path.isfile('SEQUENCE'):
+    sequence_file=os.path.join(path,'SEQUENCE')
+    if os.path.isfile(sequence_file):
         # read previous sequence UUID from the file
         print('--- SEQUENCE FILE FOUND, RESUMING UPLOAD ---')
-        with file('SEQUENCE') as f:
+        with file(sequence_file) as f:
```
Comment (ybkuang) [inline python/upload_with_authentication.py]: @richlv let's use 
`with open(sequence_file) as f` here instead for conventional reason.

Comment (ybkuang) [inline python/upload_with_authentication.py]: I will merge this after this minor tweaking from you. Thanks for fixing this and sorry for the late response.

**Comment by richlv** [comment]
@ybkuang, thanks a lot for reviewing - made the change. please note that i was not sure about style guidelines, so followed the style in other file access statements. would be glad to clean it up if there are guidelines.

**Comment by andrewharvey** [comment]
what's the status of this PR? I only saw it after trying to implement this myself over at https://github.com/mapillary/mapillary_tools/issues/168 after being frustrated by the inability to pause and resume uploads.

comparing it to the implementation at https://github.com/mapillary/mapillary_tools/issues/168, it doesn't look like sequence file is removed at the end when it's successfully completed.

**Comment by andrewharvey** [comment]
adding a ref to #57 so that this PR shows up in the issue

**Comment by ybkuang** [comment]
@richlv Thanks! 

Is it OK that you rebase your commits on top of the latest on the master, so that it becomes a single commit instead of multiple ones? 

Regarding the style, we are following [PEP8]( https://www.python.org/dev/peps/pep-0008/
)

**Comment by ybkuang** [comment]
@andrewharvey Thanks for the pull request. We're going to merge #98. Maybe you can add the sequence file removal on top of the that?

**Comment by andrewharvey** [comment]
> @andrewharvey Thanks for the pull request. We're going to merge #98. Maybe you can add the sequence file removal on top of the that?

@richlv is that something you could add to your PR? I'd rather give you the chance to address this yourself.

See how in my PR at https://github.com/mapillary/mapillary_tools/pull/168/files I remove the seq file at the end (when someone exits with Ctrl+C, I don't think it would reach this and hence leave the seq file in place).

Otherwise when running it a second time after a successful run it will always use the same seq id right right?

File: python/upload_with_authentication.py [hunk 2]
```diff
@@ -72,10 +72,18 @@
...
         sys.exit()
-
-    # generate a sequence UUID
-    sequence_id = uuid.uuid4()
-
+    sequence_file=os.path.join(path,'SEQUENCE')
+    if os.path.isfile(sequence_file):
+        # read previous sequence UUID from the file
+        print('--- SEQUENCE FILE FOUND, RESUMING UPLOAD ---')
+        with open(sequence_file,'r') as f:
```
Comment (andrewharvey) [inline python/upload_with_authentication.py]: do you need to close the file? Honestly I don't know enough of how python works, just checking.

**Comment by richlv** [comment]
@ybkuang, will look into rebasing. git always scares me :)

@andrewharvey - indeed, the sequence file is not removed the same way the "done" file is not - it was suggested to me long ago that it is better to keep those files as it gives you history

regarding closing the file, not sure - but it sounds like a good idea in general. will add that.

**Comment by andrewharvey** [comment]
@richlv @ybkuang You should be able to have github do the squash and merge, it takes away the need to try to do it yourself :wink: . Otherwise http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html should help.

PS. this feature would be a big help, at the moment every time it fails I end up reuploading from scratch. I use lepton to compress jpg locally, push to a cloud server using rsync (which deals with interrupts and restarting much better) and then finally use this upload script on the remote host to push to Mapillary. I hope you guys delete old files uploaded via these scripts otherwise there will be a lot of files sitting around never to be used.

**Comment by richlv** [comment]
@ybkuang, as @andrewharvey suggested, looks like github offers a way to squash the individual commits when merging :

> When you select the Squash and merge option on a pull request on GitHub, the pull request's commits are squashed into a single commit.

from https://help.github.com/articles/about-pull-request-merges/

is that option available to you ?

File: python/upload_with_authentication.py (hunk 2 above)
Comment (richlv) [inline python/upload_with_authentication.py]: reading on this, looks like no - "with" creates its own context, and file is closed after that context is left ("...close it as soon as you leave the context") :

https://stackoverflow.com/questions/3012488/what-is-the-python-with-statement-designed-for

**Comment by ybkuang** [comment]
@richlv @andrewharvey Thanks for the help! I will get to this as soon as possible.

**Comment by Skippern** [comment]
Any updates on progress?

**Comment by mprins** [comment]
This PR is long out of date and cannot be merged, it should be closed.

**Comment by richlv** [comment]
5 years later this indeed is out of date, as @mprins noted a few years ago.
Closing.

**Pull request closed without merging**